### DISCUSSION TITLE
I made a web server using fish and have a lot of problem to be solved.

### DISCUSSION THREAD

**Discussion by buddhi-deep** [fish-shell/fish-shell#8950] [Q&A] [answered]
[the web server](https://github.com/TeaHouseLab/darkwater)

How it works: client <-tcp-> socat <-pipe-> a script (codes/core/logicpipe.fish)

The problem is that read in fish can only read one line everytime, but http request does not consisted in only one line

Problem: I can't break the while loop at the end of request because i don't know how to identify it, use grep to filter \r\n didn't help.

```fish
while read request_raw
        if echo $request_raw | grep -qs '\r\n'
            break
       else
           set request "$request $request_raw"
        end
    end
```

**Comment by faho** [comment, marked as answer]
`read` will strip the `\n`.

That's why I told you to compare to just `\r`:

```fish
while read request_raw
    if test $request_raw = \r
       echo IS CARRIAGE RETURN
    end
end
```

works (well, technically it'll accept a \r and then an end-of-file as well).

>        if echo $request_raw | grep -qs '\r\n'

The `echo` will add a \n back, but grep doesn't understand these escapes. Try

    echo \r | grep -qs '\r\n'

That doesn't work. grep doesn't match it.

    echo \r | grep -qs \r\n

which passes an actual carriage-return and line-feed to grep does.

**Reply by buddhi-deep** [reply]
But when I test it with \r it matches not only the end line

**Reply by faho** [reply]
I don't know what you mean.

If you do `test $request_raw = \r`, then that will be true every time $request_raw is exactly `\r`, which will be the case whenever `read` found `\r\n` or `\r` and then the end of the file (i.e. the writer stopped).

This won't match e.g. `foo\r\n` because that's not equal. It has a "foo".

**Reply by buddhi-deep** [reply]
![image](https://user-images.githubusercontent.com/38799701/168019995-e5676fbe-ebb1-42be-8368-ed40b162bebf.png)

I pull out a debug string

When it match the \r it will send the line to a remote tcp server
As you can see it match not only the end line....That's wired

**Reply by faho** [reply]
That's because you're using `grep`, which defaults to matching a substring.

`grep oo` will match `foobar`.

 `test` will not. It just does a full string comparison.

If you want `grep` to match an exact string, you need to anchor it with `^` (beginning of the line) and `$` (end of the line), something like `grep '^'\r\n'$'`, or possibly use the `-x` option to grep.

**Comment by buddhi-deep** [comment]
OMG that make sense...Thank you for saving my day
